Framework/Parallel/src/IO/MultiProcessEventLoader.cpp

```diff
@@ -0,0 +1,295 @@
+    worker.join();
+}
+
+/**Wrapper for loading the PART of ("from" event "to" event) data
+ * from nexus file with different strategies*/
+void MultiProcessEventLoader::fillFromFile(
+    EventsListsShmemStorage &storage, const std::string &filename,
+    const std::string &groupname, const std::vector<std::string> &bankNames,
+    const std::vector<int32_t> &bankOffsets, unsigned from, unsigned to,
+    bool precalc) {
+  H5::H5File file(filename.c_str(), H5F_ACC_RDONLY);
+  auto instrument = file.openGroup(groupname);
+
+  auto type =
+      EventLoader::readDataType(instrument, bankNames, "event_time_offset");
+
+  if (precalc) {
+    if (type == H5::PredType::NATIVE_INT32)
+      return GroupLoader<LoadType::preCalcEvents>::loadFromGroup<int32_t>(
+          storage, instrument, bankNames, bankOffsets, from, to);
+    if (type == H5::PredType::NATIVE_INT64)
+      return GroupLoader<LoadType::preCalcEvents>::loadFromGroup<int64_t>(
+          storage, instrument, bankNames, bankOffsets, from, to);
+    if (type == H5::PredType::NATIVE_UINT32)
+      return GroupLoader<LoadType::preCalcEvents>::loadFromGroup<uint32_t>(
+          storage, instrument, bankNames, bankOffsets, from, to);
+    if (type == H5::PredType::NATIVE_UINT64)
+      return GroupLoader<LoadType::preCalcEvents>::loadFromGroup<uint64_t>(
+          storage, instrument, bankNames, bankOffsets, from, to);
+    if (type == H5::PredType::NATIVE_FLOAT)
+      return GroupLoader<LoadType::preCalcEvents>::loadFromGroup<float>(
+          storage, instrument, bankNames, bankOffsets, from, to);
+    if (type == H5::PredType::NATIVE_DOUBLE)
+      return GroupLoader<LoadType::preCalcEvents>::loadFromGroup<double>(
+          storage, instrument, bankNames, bankOffsets, from, to);
+    throw std::runtime_error(
+        "Unsupported H5::DataType for event_time_offset in NXevent_data");
+  } else {
+    if (type == H5::PredType::NATIVE_INT32)
+      return GroupLoader<LoadType::producerConsumer>::loadFromGroup<int32_t>(
+          storage, instrument, bankNames, bankOffsets, from, to);
+    if (type == H5::PredType::NATIVE_INT64)
+      return GroupLoader<LoadType::producerConsumer>::loadFromGroup<int64_t>(
+          storage, instrument, bankNames, bankOffsets, from, to);
+    if (type == H5::PredType::NATIVE_UINT32)
+      return GroupLoader<LoadType::producerConsumer>::loadFromGroup<uint32_t>(
+          storage, instrument, bankNames, bankOffsets, from, to);
+    if (type == H5::PredType::NATIVE_UINT64)
+      return GroupLoader<LoadType::producerConsumer>::loadFromGroup<uint64_t>(
+          storage, instrument, bankNames, bankOffsets, from, to);
+    if (type == H5::PredType::NATIVE_FLOAT)
+      return GroupLoader<LoadType::producerConsumer>::loadFromGroup<float>(
+          storage, instrument, bankNames, bankOffsets, from, to);
+    if (type == H5::PredType::NATIVE_DOUBLE)
+      return GroupLoader<LoadType::producerConsumer>::loadFromGroup<double>(
```

Comment:
Maybe you can move this code block to a templated function to avoid having it twice for the two load types (in case you intend to keep both).

Reply:
It's possible, but there are only 2 strategies and may be  one can be removed. If not then the proper wrapper would be introduced.